src/t8_schemes/t8_default/t8_default_common/t8_default_common.hxx

```diff
@@ -24,37 +24,179 @@
+static inline t8_gloidx_t
+count_leaves_from_level (int element_level, int refinement_level, int dimension)
+{
+  return element_level > refinement_level ? 0 : sc_intpow64 (2, dimension * (refinement_level - element_level));
+}
+
+template <class TUnderlyingEclass_Scheme>
+class t8_default_scheme_common: public t8_crtp<TUnderlyingEclass_Scheme> {
+ private:
+  friend TUnderlyingEclass_Scheme;
+  /** Private constructor which can only be used by derived schemes.
+   * \param [in] tree_class The tree class of this element scheme.
+   * \param [in] elem_size  The size of the elements this scheme holds.
+  */
+  t8_default_scheme_common (const t8_eclass_t tree_class, const size_t elem_size)
+    : element_size (elem_size), ts_context (sc_mempool_new (elem_size)), eclass (tree_class) {};
+
+ protected:
+  size_t element_size; /**< The size in bytes of an element of class \a eclass */
+  void *ts_context;    /**< Anonymous implementation context. */
+
  public:
+  t8_eclass_t eclass; /**< The tree class */
+
   /** Destructor for all default schemes */
-  virtual ~t8_default_scheme_common_c ();
+  ~t8_default_scheme_common ()
+  {
+    T8_ASSERT (ts_context != NULL);
+    SC_ASSERT (((sc_mempool_t *) ts_context)->elem_count == 0);
+    sc_mempool_destroy ((sc_mempool_t *) ts_context);
+  }
+
+  /** Move constructor */
+  t8_default_scheme_common (t8_default_scheme_common &&other) noexcept
+    : element_size (other.element_size), ts_context (other.ts_context), eclass (other.eclass)
+  {
+    other.ts_context = nullptr;
+  }
+
+  /** Move assignment operator */
+  t8_default_scheme_common &
+  operator= (t8_default_scheme_common &&other) noexcept
+  {
+    if (this != &other) {
+      // Free existing resources of moved-to object
+      if (ts_context) {
+        sc_mempool_destroy ((sc_mempool_t *) ts_context);
+      }
 
-  virtual void
-  t8_element_deinit (int length, t8_element_t *elem) const override;
+      // Transfer ownership of resources
+      element_size = other.element_size;
+      eclass = other.eclass;
+      ts_context = other.ts_context;
+
+      // Leave the source object in a valid state
+      other.ts_context = nullptr;
+    }
+    return *this;
+  }
+
+  /** Copy constructor */
+  t8_default_scheme_common (const t8_default_scheme_common &other)
+    : element_size (other.element_size), ts_context (sc_mempool_new (other.element_size)), eclass (other.eclass) {};
+
+  /** Copy assignment operator */
+  t8_default_scheme_common &
+  operator= (const t8_default_scheme_common &other)
+  {
+    if (this != &other) {
+      // Free existing resources of assigned-to object
+      if (ts_context) {
+        sc_mempool_destroy ((sc_mempool_t *) ts_context);
+      }
+
+      // Copy the values from the source object
+      element_size = other.element_size;
+      eclass = other.eclass;
+      ts_context = sc_mempool_new (other.element_size);
+    }
+    return *this;
+  }
 
   /** Compute the number of corners of a given element. */
```

Comment:
Please add a comment that pyramids have their own implementation of this function.